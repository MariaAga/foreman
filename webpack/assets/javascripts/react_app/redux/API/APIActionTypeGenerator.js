const actions = ['REQUEST', 'SUCCESS', 'FAILURE'];

/**
 * Auto generates api consts for redux for given subtype
 * @param {String} subtype - the name of the component
 * @param {Object} actionTypes - custom types to use instead of the auto generated ones
 */
export const actionTypeGenerator = (subtype, actionTypes = {}) => {
  actions.forEach(type => {
    actionTypes[type] = actionTypes[type] || `${subtype}_${type}`;
  });
  return actionTypes;
};
