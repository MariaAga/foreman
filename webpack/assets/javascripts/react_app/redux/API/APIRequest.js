import { API } from './';

export const get = async (
  payload,
  url,
  store,
  actionTypes,
  errorFormat,
  successFormat,
  onSuccess,
  onFailure
) => {
  store.dispatch({
    type: actionTypes.REQUEST,
    payload,
  });
  try {
    const response = await API.get(
      url,
      payload.headers || {},
      payload.params || {}
    );
    onSuccess(response);
    store.dispatch({
      type: actionTypes.SUCCESS,
      payload: { ...payload, ...successFormat(response.data) },
    });
  } catch (error) {
    onFailure(error);
    store.dispatch({
      type: actionTypes.FAILURE,
      payload: { ...errorFormat({ error }), payload },
    });
  }
};
