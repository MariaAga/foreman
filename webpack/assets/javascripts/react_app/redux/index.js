import createLogger from 'redux-logger';
import thunk from 'redux-thunk';
import { applyMiddleware, createStore } from 'redux';

import reducers from './reducers';

import { ajaxRequestAction } from './middlewares/ajaxRequestAction';

let middleware = [thunk, ajaxRequestAction];

const useLogger = () => {
  const isProduction = process.env.NODE_ENV === 'production';
  const isLogger = process.env.REDUX_LOGGER;

  if (!isProduction && !global.__testing__) {
    if (isLogger === undefined || isLogger === true) return true;
  }
  return isProduction && isLogger;
};

if (useLogger()) middleware = [...middleware, createLogger()];

export const generateStore = () =>
  createStore(
    reducers,
    window.__REDUX_DEVTOOLS_EXTENSION__ &&
      window.__REDUX_DEVTOOLS_EXTENSION__(),
    applyMiddleware(...middleware)
  );

const store = generateStore();

export default store;
