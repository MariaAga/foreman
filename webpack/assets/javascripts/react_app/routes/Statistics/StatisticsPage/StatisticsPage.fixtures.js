import { statisticsMeta } from '../../../components/StatisticsChartsList/StatisticsChartsList.fixtures';
import { noop } from '../../../common/helpers';

export const statisticsProps = {
  statisticsMeta,
  isLoading: false,
  hasData: true,
  hasError: false,
  message: {},
  getStatisticsMeta: noop,
};
