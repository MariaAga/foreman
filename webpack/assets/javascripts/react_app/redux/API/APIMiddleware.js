import { API_OPERATIONS, actionTypeGenerator } from './';
import { get } from './APIRequest';

export const APIMiddleware = store => next => action => {
  const { type, outputType, payload = {}, url, actionTypes = {} } = action;
  if (type === API_OPERATIONS.GET) {
    get(payload, url, store, actionTypeGenerator(outputType, actionTypes));
  } else if (type === '_START_POLLING') {
    pollingStart(
      payload.timeoutId,
      outputType,
      payload.APIParams,
      payload.interval
    );
  } else if (type === '_STOP_POLLING') {
    pollingStart(
      payload.timeoutId,
      outputType,
      payload.APIParams,
      payload.interval
    );
  } else {
    next(action);
  }
};

export const pollingStop = (timeoutId, outputType) => ({
  type: `${outputType}_STOP_POLLING`,
  payload: { timeoutId, outputType },
});

export const pollingStart = (
  timeoutId,
  outputType,
  APIParams,
  interval = 5000
) => dispatch => {
  if (!timeoutId) {
    timeoutId = setInterval(() => getData(APIParams, outputType), interval);
  }
  dispatch({
    type: `${outputType}_START_POLLING`,
    payload: { timeoutId },
  });
};

const getData = (APIParams, outputType) => ({
  type: API_OPERATIONS.GET,
  outputType,
  ...APIParams,
});
