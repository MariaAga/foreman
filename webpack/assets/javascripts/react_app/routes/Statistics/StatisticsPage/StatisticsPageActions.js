import { API_OPERATIONS } from '../../../redux/consts';

import {
  STATISTICS_PAGE_DATA_RESOLVED,
  STATISTICS_PAGE_DATA_FAILED,
  STATISTICS_PAGE_HIDE_LOADING,
  STATISTICS_PAGE_URL,
} from '../constants';

export const getStatisticsMeta = (
  url = STATISTICS_PAGE_URL
) => async dispatch => {
  const formatResults = data => ({
    metadata: data,
    hasData: data.length > 0,
  });
  const formatErrors = ({ error }) => ({
    message: {
      type: 'error',
      text: error.message,
    },
  });
  dispatch({
    type: API_OPERATIONS.GET,
    outputType: 'STATISTICS_PAGE',
    actionTypes: {
      SUCCESS: STATISTICS_PAGE_DATA_RESOLVED,
      FAILURE: STATISTICS_PAGE_DATA_FAILED,
    },
    url,
    onSuccess: () => dispatch(hideLoading()),
    onFailure: () => dispatch(hideLoading()),
    successFormat: formatResults,
    errorFormat: formatErrors,
  });
};

const hideLoading = () => ({
  type: STATISTICS_PAGE_HIDE_LOADING,
});
