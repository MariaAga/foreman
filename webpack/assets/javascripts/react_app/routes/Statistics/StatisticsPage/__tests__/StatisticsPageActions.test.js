import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import API from '../../../../redux/API/API';
import { getStatisticsMeta } from '../StatisticsPageActions';
import { statisticsProps } from '../StatisticsPage.fixtures';

import { APIMiddleware } from '../../../../redux/API';
import IntegrationTestHelper from '../../../../common/IntegrationTestHelper';

const middlewares = [thunk, APIMiddleware];
const mockStore = configureMockStore(middlewares);
const store = mockStore();

afterEach(() => {
  store.clearActions();
});

jest.mock('../../../../redux/API/API');

describe('StatisticsPage actions', () => {
  it('should fetch statisticsMeta', async () => {
    API.get.mockImplementation(
      () =>
        new Promise((resolve, reject) => {
          resolve({
            data: statisticsProps.statisticsMeta,
          });
        })
    );
    await store.dispatch(getStatisticsMeta());
    await IntegrationTestHelper.flushAllPromises();
    expect(store.getActions()).toMatchSnapshot();
  });
  it('should fetch statisticsMeta and fail', async () => {
    API.get.mockImplementation(
      () =>
        new Promise((resolve, reject) => {
          reject(Error('some-error'));
        })
    );
    await store.dispatch(getStatisticsMeta());
    await IntegrationTestHelper.flushAllPromises();
    expect(store.getActions()).toMatchSnapshot();
  });
});
