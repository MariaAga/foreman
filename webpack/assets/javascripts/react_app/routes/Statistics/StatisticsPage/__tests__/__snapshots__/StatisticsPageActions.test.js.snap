// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsPage actions should fetch statisticsMeta 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "STATISTICS_PAGE_REQUEST",
  },
  Object {
    "type": "STATISTICS_PAGE_HIDE_LOADING",
  },
  Object {
    "payload": Object {
      "hasData": true,
      "metadata": Array [
        Object {
          "id": "operatingsystem",
          "search": "/hosts?search=os_title=~VAL~",
          "title": "OS Distribution",
          "url": "statistics/operatingsystem",
        },
        Object {
          "id": "architecture",
          "search": "/hosts?search=facts.architecture=~VAL~",
          "title": "Architecture Distribution",
          "url": "statistics/architecture",
        },
      ],
    },
    "type": "STATISTICS_PAGE_DATA_RESOLVED",
  },
]
`;

exports[`StatisticsPage actions should fetch statisticsMeta and fail 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "STATISTICS_PAGE_REQUEST",
  },
  Object {
    "type": "STATISTICS_PAGE_HIDE_LOADING",
  },
  Object {
    "payload": Object {
      "message": Object {
        "text": "some-error",
        "type": "error",
      },
      "payload": Object {},
    },
    "type": "STATISTICS_PAGE_FETCH_FAILED",
  },
]
`;
